containers/src/Data/IntMap/Internal.hs

```diff
@@ -62,2904 +275,1189 @@
+instance NFData a => NFData (IntMap a) where
+    rnf (IntMap Empty) = ()
+    rnf (IntMap (NonEmpty _ v n)) = rnf v `seq` rnf n
 
--- This instance preserves data abstraction at the cost of inefficiency.
--- We provide limited reflection services for the sake of data abstraction.
+instance NFData a => NFData (Node t a) where
+    rnf Tip = ()
+    rnf (Bin _ v l r) = rnf v `seq` rnf l `seq` rnf r
 
-instance Data a => Data (IntMap a) where
-  gfoldl f z im = z fromList `f` (toList im)
-  toConstr _     = fromListConstr
-  gunfold k z c  = case constrIndex c of
-    1 -> k (z fromList)
-    _ -> error "gunfold"
-  dataTypeOf _   = intMapDataType
-  dataCast1 f    = gcast1 f
+-- | /O(min(n,W))/. Find the value at a key.
+-- Calls 'error' when the element can not be found.
+--
+-- > fromList [(5,'a'), (3,'b')] ! 1    Error: element not in the map
+-- > fromList [(5,'a'), (3,'b')] ! 5 == 'a'
+(!) :: IntMap a -> Key -> a
+(!) m k = findWithDefault (error $ "IntMap.!: key " ++ show k ++ " is not an element of the map") k m
 
-fromListConstr :: Constr
-fromListConstr = mkConstr intMapDataType "fromList" [] Prefix
 
-intMapDataType :: DataType
-intMapDataType = mkDataType "Data.IntMap.Internal.IntMap" [fromListConstr]
+-- | /O(min(n,W))/. Find the value at a key.
+-- Returns 'Nothing' when the element can not be found.
+--
+-- > fromList [(5,'a'), (3,'b')] !? 1 == Nothing
+-- > fromList [(5,'a'), (3,'b')] !? 5 == Just 'a'
+--
+-- @since 0.5.11
+(!?) :: IntMap a -> Key -> Maybe a
+(!?) m k = lookup k m
 
-#endif
+-- | Same as 'difference'.
+(\\) :: IntMap a -> IntMap b -> IntMap a
+(\\) = difference
 
-{--------------------------------------------------------------------
-  Query
---------------------------------------------------------------------}
 -- | /O(1)/. Is the map empty?
 --
--- > Data.IntMap.null (empty)           == True
+-- > Data.IntMap.null empty             == True
 -- > Data.IntMap.null (singleton 1 'a') == False
-
 null :: IntMap a -> Bool
-null Nil = True
-null _   = False
-{-# INLINE null #-}
+null (IntMap Empty) = True
+null _ = False
 
 -- | /O(n)/. Number of elements in the map.
 --
 -- > size empty                                   == 0
 -- > size (singleton 1 'a')                       == 1
 -- > size (fromList([(1,'a'), (2,'c'), (3,'b')])) == 3
 size :: IntMap a -> Int
-size = go 0
-  where
-    go !acc (Bin _ _ l r) = go (go acc l) r
-    go acc (Tip _ _) = 1 + acc
-    go acc Nil = acc
+size (IntMap Empty) = 0
+size (IntMap (NonEmpty _ _ node)) = sizeNode node where
+    sizeNode :: Node t a -> Int
+    sizeNode Tip = 1
+    sizeNode (Bin _ _ l r) = sizeNode l + sizeNode r
 
 -- | /O(min(n,W))/. Is the key a member of the map?
 --
 -- > member 5 (fromList [(5,'a'), (3,'b')]) == True
 -- > member 1 (fromList [(5,'a'), (3,'b')]) == False
-
--- See Note: Local 'go' functions and capturing]
 member :: Key -> IntMap a -> Bool
-member !k = go
+member k = k `seq` start
```

Comment:
Why not use a bang pattern for `k`?

Reply:
It's also pretty weird to start doing work before getting the second argument. Just use `member !k m0 = start m0`. Or it might be faster to use `member k m0 = start k m0`, passing the key unchanged through the recursion (only benchmarking will determine that).

Reply:
I think I switched to using bang patterns partway through writing this code. I'll clean up everything to use bang patterns where possible.